File: android/app/src/main/java/com/desmoines/aipulse/ui/screens/discover/DiscoverViewModel.kt
package com.desmoines.aipulse.ui.screens.discover

import androidx.compose.runtime.Immutable
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.desmoines.aipulse.data.model.PaywallContext
import com.desmoines.aipulse.data.model.SwipeItem
import com.desmoines.aipulse.data.model.SwipeItemType
import com.desmoines.aipulse.data.remote.EventsQuery
import com.desmoines.aipulse.data.remote.FavoritesException
import com.desmoines.aipulse.data.remote.RestaurantsQuery
import com.desmoines.aipulse.data.repository.AuthRepository
import com.desmoines.aipulse.data.repository.EventsRepository
import com.desmoines.aipulse.data.repository.FavoritesRepository
import com.desmoines.aipulse.data.repository.RestaurantsRepository
import com.desmoines.aipulse.util.SoftPaywallService
import com.desmoines.aipulse.util.SwipeInteractionService
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.async
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import java.time.format.DateTimeFormatter
import javax.inject.Inject

@Immutable
data class DiscoverUiState(
    val items: List<SwipeItem> = emptyList(),
    val isLoading: Boolean = true,
    val errorMessage: String? = null,
    /** Number of items saved (liked) this session — drives the saved-count badge. */
    val savedCount: Int = 0,
    val mode: DiscoverMode = DiscoverMode.MIXED,
    /** Filter chips shown above the deck (boost narrowing or a typed pre-filter). */
    val filterTags: List<String> = emptyList(),
    /** True when a typed pre-filter locks the mode toggle. */
    val modeLocked: Boolean = false,
) {
    val isEmpty: Boolean get() = items.isEmpty() && !isLoading && errorMessage == null
}

/**
 * Feeds + scopes the Discover swipe deck (ANDP-020/021/022). Owns the mode
 * toggle, filter context, independent paged fetches with a stale-result
 * generation guard, background prefetch, and "more like this" boost narrowing —
 * plus swipe persistence (records signals, dedupes, likes → favorites).
 * Mirrors iOS DiscoverViewModel.
 */
@HiltViewModel
class DiscoverViewModel @Inject constructor(
    private val eventsRepository: EventsRepository,
    private val restaurantsRepository: RestaurantsRepository,
    private val favoritesRepository: FavoritesRepository,
    private val authRepository: AuthRepository,
    private val swipeInteractionService: SwipeInteractionService,
    private val softPaywallService: SoftPaywallService,
) : ViewModel() {

    private val _uiState = MutableStateFlow(DiscoverUiState())
    val uiState: StateFlow<DiscoverUiState> = _uiState.asStateFlow()

    private var filter = DiscoverFilterContext()

    private var favoriteEventIds: Set<String> = emptySet()
    private var favoriteRestaurantIds: Set<String> = emptySet()

    // Paging + stale-result guard.
    private var eventOffset = 0
    private var restaurantOffset = 0
    private var hasMoreEvents = true
    private var hasMoreRestaurants = true
    private var fetchGeneration = 0
    private var fetchErrored = false
    private val fetchMutex = Mutex()

    init {
        loadFavorites()
        reload()
    }

    private fun loadFavorites() {
        val userId = authRepository.currentUserId ?: return
        viewModelScope.launch {
            favoritesRepository.loadFavorites(userId).onSuccess { state ->
                favoriteEventIds = state.favoriteEventIds
                favoriteRestaurantIds = state.favoriteRestaurantIds
            }
        }
    }

    fun setMode(mode: DiscoverMode) {
        if (mode == _uiState.value.mode || _uiState.value.modeLocked) return
        _uiState.update { it.copy(mode = mode) }
        reload()
    }

    fun clearFilter() {
        if (filter.isEmpty || _uiState.value.modeLocked) return
        filter = DiscoverFilterContext()
        reload()
    }

    fun reload() {
        _uiState.update { it.copy(isLoading = true, items = emptyList(), errorMessage = null, filterTags = filter.tags()) }
        resetPaging()
        fetchGeneration++
        viewModelScope.launch {
            fetchMore()
            _uiState.update {
                it.copy(
                    isLoading = false,
                    errorMessage = if (it.items.isEmpty() && fetchErrored) {
                        "Couldn't load the discovery deck. Try again."
                    } else null,
                )
            }
        }
    }

    fun onLike(item: SwipeItem) {
        record(SwipeInteractionService.Action.LIKE, item)
        advancePast(item)
        saveFavorite(item)
    }

    fun onSkip(item: SwipeItem) {
        record(SwipeInteractionService.Action.SKIP, item)
        advancePast(item)
    }

    /** Boost = strong positive + "show me more like this": narrow the filter and refetch. */
    fun onBoost(item: SwipeItem) {
        record(SwipeInteractionService.Action.BOOST, item)
        applyBoostFilter(item)
        reload()
    }

    /** Detail tap records a weak-positive signal; navigation is handled by the caller. */
    fun onDetail(item: SwipeItem) {
        record(SwipeInteractionService.Action.DETAIL, item)
    }

    fun clearError() = _uiState.update { it.copy(errorMessage = null) }

    // MARK: - Internals

    private fun applyBoostFilter(item: SwipeItem) {
        filter = when (item.itemType) {
            SwipeItemType.EVENT -> filter.copy(eventCategory = item.eventCategory ?: filter.eventCategory)
            SwipeItemType.RESTAURANT ->
                item.cuisine?.let { filter.copy(cuisines = listOf(it)) } ?: filter
        }
    }

    private fun record(action: SwipeInteractionService.Action, item: SwipeItem) {
        viewModelScope.launch {
            swipeInteractionService.record(action, itemTypeString(item), item.rawId, filter.toSourceContext())
        }
    }

    private fun saveFavorite(item: SwipeItem) {
        val userId = authRepository.currentUserId ?: return
        viewModelScope.launch {
            val result = when (item.itemType) {
                SwipeItemType.EVENT ->
                    favoritesRepository.toggleEventFavorite(userId, item.rawId, favoriteEventIds)
                SwipeItemType.RESTAURANT ->
                    favoritesRepository.toggleRestaurantFavorite(userId, item.rawId, favoriteRestaurantIds)
            }
            result
                .onSuccess { favorited ->
                    if (favorited) {
                        when (item.itemType) {
                            SwipeItemType.EVENT -> favoriteEventIds = favoriteEventIds + item.rawId
                            SwipeItemType.RESTAURANT -> favoriteRestaurantIds = favoriteRestaurantIds + item.rawId
                        }
                        _uiState.update { it.copy(savedCount = it.savedCount + 1) }
                    }
                }
                .onFailure { error ->
                    // At the free cap, soft-fail to the paywall rather than show a
                    // Discover-specific error; other failures are logged, not surfaced.
                    if (error is FavoritesException.LimitReached) {
                        softPaywallService.maybePresent(PaywallContext.FAVORITES)
                    }
                }
        }
    }

    /** Remove the committed card; prefetch the next page when the deck runs low. */
    private fun advancePast(item: SwipeItem) {
        _uiState.update { state -> state.copy(items = state.items.filterNot { it.id == item.id }) }
        if (_uiState.value.items.size <= PREFETCH_THRESHOLD && !fetchMutex.isLocked) {
            viewModelScope.launch { fetchMore() }
        }
    }

    private fun resetPaging() {
        eventOffset = 0
        restaurantOffset = 0
        hasMoreEvents = true
        hasMoreRestaurants = true
    }

    private suspend fun fetchMore() {
        fetchMutex.withLock {
            fetchErrored = false
            val generation = fetchGeneration
            when (_uiState.value.mode) {
                DiscoverMode.EVENTS -> fetchEventBatch(generation)
                DiscoverMode.RESTAURANTS -> fetchRestaurantBatch(generation)
                DiscoverMode.MIXED -> coroutineScope {
                    val events = async { fetchEventBatch(generation) }
                    val restaurants = async { fetchRestaurantBatch(generation) }
                    events.await()
                    restaurants.await()
                }
            }
        }
    }

    private suspend fun fetchEventBatch(generation: Int) {
        if (!hasMoreEvents) return
        // "Tonight" (mixed) with no explicit preset shows the next 7 days so the
        // deck isn't empty, without pulling events weeks out.
        val preset = filter.datePreset
            ?: if (_uiState.value.mode == DiscoverMode.MIXED) com.desmoines.aipulse.data.model.DateFilterPreset.THIS_WEEK else null
        val range = preset?.dateRange
        val query = EventsQuery(
            category = filter.eventCategory,
            dateStart = range?.first?.format(ISO),
            dateEnd = range?.second?.format(ISO),
            limit = PAGE_SIZE,
            offset = eventOffset,
        )
        eventsRepository.fetchEvents(query)
            .onSuccess { response ->
                if (generation != fetchGeneration) return
                val fresh = response.events
                    .filterNot { swipeInteractionService.hasSwiped("event", it.id) }
                    .map(SwipeItem::from)
                appendUnique(fresh)
                eventOffset += response.events.size
                hasMoreEvents = response.hasMore
            }
            .onFailure {
                if (generation != fetchGeneration) return
                hasMoreEvents = false
                fetchErrored = true
            }
    }

    private suspend fun fetchRestaurantBatch(generation: Int) {
        if (!hasMoreRestaurants) return
        val query = RestaurantsQuery(
            cuisines = filter.cuisines.ifEmpty { null },
            priceRanges = filter.priceRanges.ifEmpty { null },
            locations = filter.locations.ifEmpty { null },
            minRating = filter.minRating.takeIf { it > 0 },
            limit = PAGE_SIZE,
            offset = restaurantOffset,
        )
        restaurantsRepository.fetchRestaurants(query)
            .onSuccess { response ->
                if (generation != fetchGeneration) return
                val fresh = response.restaurants
                    .let { list -> if (filter.openNow) list.filter { it.isOpenNow() == true } else list }
                    .filterNot { swipeInteractionService.hasSwiped("restaurant", it.id) }
                    .map(SwipeItem::from)
                appendUnique(fresh)
                restaurantOffset += response.restaurants.size
                hasMoreRestaurants = response.hasMore
            }
            .onFailure {
                if (generation != fetchGeneration) return
                hasMoreRestaurants = false
                fetchErrored = true
            }
    }

    private fun appendUnique(newItems: List<SwipeItem>) {
        _uiState.update { state ->
            val existing = state.items.mapTo(HashSet()) { it.id }
            val unique = newItems.filterNot { existing.contains(it.id) }
            // Mixed mode: shuffle the new batch so events and restaurants interleave
            // instead of arriving in two solid blocks.
            val toAppend = if (state.mode == DiscoverMode.MIXED) unique.shuffled() else unique
            state.copy(items = state.items + toAppend)
        }
    }

    private fun itemTypeString(item: SwipeItem): String = when (item.itemType) {
        SwipeItemType.EVENT -> "event"
        SwipeItemType.RESTAURANT -> "restaurant"
    }

    companion object {
        private const val PAGE_SIZE = 20
        private const val PREFETCH_THRESHOLD = 4
        private val ISO: DateTimeFormatter = DateTimeFormatter.ISO_LOCAL_DATE_TIME
    }
}

File: ios/DesMoinesInsider/ViewModels/DiscoverViewModel.swift
import Foundation

// MARK: - Mode

enum DiscoverMode: String, CaseIterable, Identifiable {
    case mixed
    case events
    case restaurants

    var id: String { rawValue }

    var title: String {
        switch self {
        case .mixed: return "Tonight"
        case .events: return "Events"
        case .restaurants: return "Dining"
        }
    }

    var icon: String {
        switch self {
        case .mixed: return "sparkles"
        case .events: return "calendar"
        case .restaurants: return "fork.knife"
        }
    }
}

// MARK: - Item wrapper

/// Type-erased deck item. Lets a single `SwipeCard` view render either an
/// Event or a Restaurant without the card itself knowing about either model.
enum SwipeItem: Identifiable, Hashable {
    case event(Event)
    case restaurant(Restaurant)

    var id: String {
        switch self {
        case .event(let e): return "event-\(e.id)"
        case .restaurant(let r): return "restaurant-\(r.id)"
        }
    }

    var imageUrl: String? {
        switch self {
        case .event(let e): return e.imageUrl
        case .restaurant(let r): return r.imageUrl
        }
    }

    var title: String {
        switch self {
        case .event(let e): return e.title
        case .restaurant(let r): return r.name
        }
    }

    var subtitle: String {
        switch self {
        case .event(let e):
            if let d = e.parsedDate {
                return d.formatted(.dateTime.weekday(.abbreviated).month(.abbreviated).day())
            }
            return e.eventCategory.displayName
        case .restaurant(let r):
            return [r.cuisine, r.priceRange].compactMap { $0 }.joined(separator: " · ")
        }
    }

    /// Short location label for the swipe card (city / venue only — not the
    /// full geocoded address string which can be 80+ characters long and
    /// causes the HStack subtitle row to show mid-string content).
    var locationText: String {
        switch self {
        case .event(let e):
            // Events already store a clean [venue, city] display location.
            return e.displayLocation
        case .restaurant(let r):
            // Use only the city; the raw `location` column contains the full
            // geocoded string ("1234 Main St, Des Moines, IA 50309, USA") which
            // is far too long for a one-line card label.
            return r.city ?? ""
        }
    }

    var typeIcon: String {
        switch self {
        case .event: return "calendar"
        case .restaurant: return "fork.knife"
        }
    }

    var typeLabel: String {
        switch self {
        case .event(let e): return e.eventCategory.displayName
        case .restaurant(let r): return r.cuisine ?? "Restaurant"
        }
    }

    var supabaseItemType: SwipeInteractionService.ItemType {
        switch self {
        case .event: return .event
        case .restaurant: return .restaurant
        }
    }

    var rawId: String {
        switch self {
        case .event(let e): return e.id
        case .restaurant(let r): return r.id
        }
    }
}

// MARK: - Filter context

/// Optional filter context that callers can hand to DiscoverView so the deck
/// is pre-narrowed (e.g. "Italian restaurants in East Village"). When all
/// fields are empty the deck pulls from the full catalog.
struct DiscoverFilterContext: Equatable {
    var cuisines: [String] = []
    var locations: [String] = []
    var priceRanges: [String] = []
    var minRating: Double = 0
    var openNow: Bool = false

    var eventCategory: EventCategory? = nil
    var datePreset: DateFilterPreset? = nil
    var freeOnly: Bool = false

    var isEmpty: Bool {
        cuisines.isEmpty && locations.isEmpty && priceRanges.isEmpty &&
        minRating == 0 && !openNow &&
        eventCategory == nil && datePreset == nil && !freeOnly
    }

    /// JSON snapshot stored alongside each swipe to preserve the user's
    /// filter context at the time of the gesture.
    func toSourceContext() -> [String: [String]]? {
        var ctx: [String: [String]] = [:]
        if !cuisines.isEmpty { ctx["cuisines"] = cuisines }
        if !locations.isEmpty { ctx["locations"] = locations }
        if !priceRanges.isEmpty { ctx["priceRanges"] = priceRanges }
        if let cat = eventCategory { ctx["category"] = [cat.rawValue] }
        if let preset = datePreset { ctx["datePreset"] = [preset.rawValue] }
        if openNow { ctx["openNow"] = ["true"] }
        if freeOnly { ctx["freeOnly"] = ["true"] }
        if minRating > 0 { ctx["minRating"] = [String(minRating)] }
        return ctx.isEmpty ? nil : ctx
    }
}

// MARK: - ViewModel

@MainActor
@Observable
final class DiscoverViewModel {
    var mode: DiscoverMode {
        didSet { if oldValue != mode { Task { await reload() } } }
    }
    private(set) var filter: DiscoverFilterContext
    private(set) var deck: [SwipeItem] = []

    /// A liked card could not be saved for a reason that is not the free-tier
    /// cap. The cap has its own app-level paywall, so surfacing it here too
    /// would double up (IOS-AUDIT-UX-057).
    private(set) var favoriteSaveFailed = false

    func acknowledgeFavoriteFailure() {
        favoriteSaveFailed = false
    }
    private(set) var isLoading = false

    /// True when the last batch fetch threw (IOS-AUDIT-UX-051 AC3).
    ///
    /// Both batch fetchers used to swallow their error and set hasMore... = false,
    /// so a network failure produced an empty deck and the "You've seen
    /// everything" screen. That tells the user they have exhausted the content
    /// when in fact nothing loaded - and the only affordance offered was a Reset
    /// that would fail the same way, silently.
    private(set) var lastLoadFailed = false
    private(set) var totalSwipes = 0
    private(set) var likedItems: [SwipeItem] = []

    /// Number of cards remaining when we trigger a background prefetch.
    private let prefetchThreshold = 4
    private let pageSize = 20

    private var eventOffset = 0
    private var restaurantOffset = 0
    private var hasMoreEvents = true
    private var hasMoreRestaurants = true
    private var isPrefetching = false

    /// Bumped whenever the deck is reset (reload / boost). A fetch captures the
    /// generation when it runs and discards its results if the generation has
    /// since changed — so an in-flight pre-reset fetch can't append stale,
    /// wrong-filter cards into the freshly-reset deck.
    private var fetchGeneration = 0
    /// The most recently enqueued fetch. New fetches chain after it so a reset's
    /// refetch is never skipped by the `isPrefetching` guard while an older
    /// fetch is still draining.
    private var fetchTask: Task<Void, Never>?

    private let eventsService: EventPageProviding
    private let restaurantsService: RestaurantPageProviding
    private let swipeService = SwipeInteractionService.shared

    /// Providers default to the shared services, so no call site changes. They
    /// exist so a test can hold a fetch open and bump the deck generation while
    /// it is in flight - the only way the discard behaviour this view model
    /// depends on can be observed (IOS-AUDIT-TEST-006).
    init(
        mode: DiscoverMode = .mixed,
        filter: DiscoverFilterContext = .init(),
        eventsService: EventPageProviding = EventsService.shared,
        restaurantsService: RestaurantPageProviding = RestaurantsService.shared
    ) {
        self.mode = mode
        self.filter = filter
        self.eventsService = eventsService
        self.restaurantsService = restaurantsService
    }

    // MARK: - Load

    func loadInitial() async {
        guard deck.isEmpty else { return }
        await reload()
    }

    func reload() async {
        isLoading = true
        lastLoadFailed = false
        deck = []
        eventOffset = 0
        restaurantOffset = 0
        hasMoreEvents = true
        hasMoreRestaurants = true
        // Invalidate in-flight fetches and run the fresh fetch after they drain
        // (so the isPrefetching guard can't skip it).
        fetchGeneration += 1
        await enqueueFetch().value
        isLoading = false
    }

    func updateFilter(_ newFilter: DiscoverFilterContext) async {
        guard newFilter != filter else { return }
        filter = newFilter
        await reload()
    }

    // MARK: - Swipe handling
    //
    // Deck mutation runs synchronously so SwiftUI sees the model update in
    // the same render frame the swipe gesture commits — otherwise the
    // newly-promoted top card briefly inherits the off-screen drag offset
    // and flickers. Persistence + signal recording are dispatched as
    // fire-and-forget Tasks.

    /// Right swipe → like. Records the signal and adds to favorites.
    func like(_ item: SwipeItem) {
        likedItems.append(item)
        advance()
        Task { await self.record(.like, item: item) }
        Task { await self.persistFavorite(item) }
    }

    /// Left swipe → skip. Pure negative signal.
    func skip(_ item: SwipeItem) {
        advance()
        Task { await self.record(.skip, item: item) }
    }

    /// Up swipe → boost. Strong positive signal *and* re-shapes the deck so
    /// the next batch leans toward the same category / cuisine. This is the
    /// "show me more like this" lever — the reason the up-swipe exists.
    func boost(_ item: SwipeItem) {
        applyBoostFilter(from: item)
        // Drop everything currently in the deck — the user just told us
        // they want a different slice — and refetch with the narrowed
        // filter on a background task.
        deck = []
        eventOffset = 0
        restaurantOffset = 0
        hasMoreEvents = true
        hasMoreRestaurants = true
        totalSwipes += 1
        // Invalidate any in-flight (pre-boost) fetch so its results are dropped
        // instead of mixed into the narrowed deck.
        fetchGeneration += 1
        // Show the loading state while the narrowed batch loads instead of
        // flashing the "you've seen everything" empty state, which the deck-
        // empty branch would otherwise render mid-boost (IOS-AUDIT-UX-019).
        isLoading = true
        Task { await self.record(.boost, item: item) }
        let task = enqueueFetch()
        Task {
            await task.value
            self.isLoading = false
        }
    }

    /// Tap → opened detail view. Logged as a positive but weaker signal.
    func recordDetailTap(_ item: SwipeItem) {
        Task { await self.record(.detail, item: item) }
    }

    // MARK: - Internals

    private func advance() {
        totalSwipes += 1
        // Drop the top card. SwipeCardStack also removes it visually; this
        // call is the single source of truth for the data model.
        if !deck.isEmpty { deck.removeFirst() }
        // Throttle background prefetch: only enqueue when nothing is already
        // fetching, so rapid swipes don't queue N sequential page fetches.
        // (reload/boost call enqueueFetch unconditionally — they must always run.)
        if deck.count <= prefetchThreshold && !isPrefetching { enqueueFetch() }
    }

    private func record(_ action: SwipeInteractionService.Action, item: SwipeItem) async {
        await swipeService.record(
            action: action,
            itemType: item.supabaseItemType,
            itemId: item.rawId,
            sourceContext: filter.toSourceContext()
        )
    }

    private func persistFavorite(_ item: SwipeItem) async {
        do {
            switch item {
            case .event(let e):
                if !FavoritesService.shared.isEventFavorited(e.id) {
                    _ = try await FavoritesService.shared.toggleFavorite(eventId: e.id)
                }
            case .restaurant(let r):
                if !FavoritesService.shared.isRestaurantFavorited(r.id) {
                    _ = try await FavoritesService.shared.toggleRestaurantFavorite(restaurantId: r.id)
                }
            }
        } catch {
            // THE CAP IS ALREADY HANDLED, and not by this view. enforceFavoritesCap
            // posts .favoritesLimitReached, which MainTabView turns into the
            // app-level upsell paywall - so a toast here would be a second,
            // redundant message on top of it. That is what
            // FavoritesService.isLimitReached exists for.
            //
            // EVERYTHING ELSE was swallowed with it: a dropped connection or an
            // expired session meant the card animated away as a save and nothing
            // was saved, with no output of any kind. The previous comment claimed
            // DiscoverView surfaced this via toast; DiscoverView declares a toast
            // and assigns it nowhere (IOS-AUDIT-UX-057).
            if !FavoritesService.isLimitReached(error) {
                favoriteSaveFailed = true
            }
        }
    }

    private func applyBoostFilter(from item: SwipeItem) {
        switch item {
        case .event(let e):
            filter.eventCategory = e.eventCategory
        case .restaurant(let r):
            if let cuisine = r.cuisine, !cuisine.isEmpty {
                filter.cuisines = [cuisine]
            }
        }
    }

    // MARK: - Fetch

    /// Enqueue a fetch that runs after any in-flight fetch drains, so a reset
    /// (reload/boost) isn't skipped by the `isPrefetching` guard. Returns the
    /// task so callers can await completion.
    @discardableResult
    private func enqueueFetch() -> Task<Void, Never> {
        let previous = fetchTask
        let task = Task {
            await previous?.value
            await self.fetchMore()
        }
        fetchTask = task
        return task
    }

    private func fetchMore() async {
        guard !isPrefetching else { return }
        isPrefetching = true
        defer { isPrefetching = false }

        // Capture the deck generation for this run; the batch fetchers drop their
        // results if a reset bumped it while we were awaiting the network.
        let generation = fetchGeneration

        switch mode {
        case .events:
            await fetchEventBatch(generation)
        case .restaurants:
            await fetchRestaurantBatch(generation)
        case .mixed:
            async let events: () = fetchEventBatch(generation)
            async let restaurants: () = fetchRestaurantBatch(generation)
            _ = await (events, restaurants)
        }
    }

    private func fetchEventBatch(_ generation: Int) async {
        guard hasMoreEvents else { return }
        var query = EventsService.EventsQuery()
        query.category = filter.eventCategory?.rawValue
        query.cities = filter.locations.isEmpty ? nil : filter.locations
        query.freeOnly = filter.freeOnly
        query.limit = pageSize
        query.offset = eventOffset
        if let preset = filter.datePreset {
            let range = preset.dateRange
            query.dateStart = range.start
            query.dateEnd = range.end
        } else if mode == .mixed {
            // "Tonight" tab: show events within the next 7 days so the deck
            // isn't empty on nights with few events, but doesn't pull events
            // from weeks/months away.
            let range = DateFilterPreset.thisWeek.dateRange
            query.dateStart = range.start
            query.dateEnd = range.end
        }

        do {
            let response = try await eventsService.fetchEvents(query: query)
            // A reload/boost reset the deck while we were awaiting — these results
            // belong to the old filter/offset; drop them rather than mixing in.
            guard generation == fetchGeneration else { return }
            let fresh = response.events
                .filter { !swipeService.hasSwiped(itemType: .event, itemId: $0.id) }
                .map { SwipeItem.event($0) }
            appendUnique(fresh)
            eventOffset += response.events.count
            hasMoreEvents = response.hasMore
        } catch {
            guard generation == fetchGeneration else { return }
            hasMoreEvents = false
            lastLoadFailed = true
        }
    }

    private func fetchRestaurantBatch(_ generation: Int) async {
        guard hasMoreRestaurants else { return }
        var query = RestaurantsService.RestaurantsQuery()
        query.cuisines = filter.cuisines.isEmpty ? nil : filter.cuisines
        query.locations = filter.locations.isEmpty ? nil : filter.locations
        query.priceRanges = filter.priceRanges.isEmpty ? nil : filter.priceRanges
        query.minRating = filter.minRating > 0 ? filter.minRating : nil
        query.sortBy = .popularity
        query.limit = pageSize
        query.offset = restaurantOffset

        do {
            let response = try await restaurantsService.fetchRestaurants(query: query)
            // Drop results if a reload/boost reset the deck mid-flight.
            guard generation == fetchGeneration else { return }
            var fresh = response.restaurants
            if filter.openNow {
                fresh = fresh.filter { $0.isOpenNow() == true }
            }
            let mapped = fresh
                .filter { !swipeService.hasSwiped(itemType: .restaurant, itemId: $0.id) }
                .map { SwipeItem.restaurant($0) }
            appendUnique(mapped)
            restaurantOffset += response.restaurants.count
            hasMoreRestaurants = response.hasMore
        } catch {
            guard generation == fetchGeneration else { return }
            hasMoreRestaurants = false
            lastLoadFailed = true
        }
    }

    private func appendUnique(_ items: [SwipeItem]) {
        let existing = Set(deck.map(\.id))
        let unique = items.filter { !existing.contains($0.id) }
        // For mixed mode, shuffle the new batch so events and restaurants
        // interleave instead of arriving in two solid blocks.
        deck.append(contentsOf: mode == .mixed ? unique.shuffled() : unique)
    }
}
